### PR TITLE
Add an API compatibility policy.

### PR DESCRIPTION
**Pull request by hawkinsp**

### CODE REVIEW

File: docs/api_compatibility.md [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+* `jax.lax`
+* `jax.nn`
+* `jax.numpy`
+* `jax.ops`
+* `jax.profiler`
+* `jax.random`
+* `jax.scipy`
+* `jax.tree_util`
+
+Not everything in these modules is public. Over time, we are working to separate
```
Comment (yashk2810) [inline docs/api_compatibility.md:31]: Maybe move this after ## What is not covered list?

File: docs/api_compatibility.md [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+public and private APIs.
```
Comment (jakevdp) [inline docs/api_compatibility.md:32]: Mention that our aim is to follow the convention of hiding all non-public apis behind names that begin with an underscore (although we're not entirely compliant with that yet)

File: docs/api_compatibility.md (hunk 1 above)
Comment (hawkinsp) [reply docs/api_compatibility.md:31]: This statement applies to the preceding list.

**Review by mattjj — APPROVED** [review]
Thanks for doing this. SGTM. I wonder if someone with experience in OSS deprecation policies could give any feedback, like @shoyer .

File: docs/api_compatibility.md (hunk 2 above)
Comment (shoyer) [reply docs/api_compatibility.md:32]: I would also note that public APIs are documented in the JAX API docs

File: docs/api_compatibility.md [hunk 3]
```diff
@@ -0,0 +1,47 @@
+# API compatibility
+
+JAX is constantly evolving, and we are not yet ready to commit to a stable API.
+That said, we want to minimize churn for the JAX user community.
+
+JAX follows a 3 month deprecation policy. When an incompatible change is made
```
Comment (shoyer) [inline docs/api_compatibility.md:6]: 3 months is fast for an OSS project. The challenge is that you need to wait for downstream libraries to release fixes, and that can take time. It's not just about giving end-users time to upgrade.

6 months would be more in line with general norms. Definitely up to you, though, there are many tradeoffs here.

File: docs/api_compatibility.md [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+to an API, we will make our best effort to obey the following procedure:
+* the change will be announced in `CHANGELOG.md`, and the old API will issue a
+  `DeprecationWarning`.
```
Comment (shoyer) [inline docs/api_compatibility.md:9]: It's worth noting that DeprecationWarning is often silenced by default!

FutureWarning is a better choice for something you want end-users to see.

File: docs/api_compatibility.md [hunk 5]
```diff
@@ -0,0 +1,47 @@
+# API compatibility
+
+JAX is constantly evolving, and we are not yet ready to commit to a stable API.
+That said, we want to minimize churn for the JAX user community.
+
+JAX follows a 3 month deprecation policy. When an incompatible change is made
+to an API, we will make our best effort to obey the following procedure:
+* the change will be announced in `CHANGELOG.md`, and the old API will issue a
```
Comment (shoyer) [inline docs/api_compatibility.md:8]: Also in the docstring for the deprecated API

File: docs/api_compatibility.md [hunk 6]
```diff
@@ -0,0 +1,47 @@
+# API compatibility
+
+JAX is constantly evolving, and we are not yet ready to commit to a stable API.
```
Comment (gnecula) [inline docs/api_compatibility.md:3]: This statement looks scarier than the reality. In reality, we remove APIs very rarely, and we think carefully before reducing functionality or making semantic changes. One can get a sense of the frequency of such changes by scanning the CHANGELOG. I think that this is worth pointing out.

I would also rewrite "not yet ready to commit to a stable API" to something like "and we want to be able to make improvements to the APIs".

File: docs/api_compatibility.md (hunk 3 above)
Comment (gnecula) [reply docs/api_compatibility.md:6]: I also think that 3 months is fast. I expect that a lower bound on that interval will be our ability to migrate internal Google users off an API end point. Most likely a widely used API point will take much longer to remove.

File: docs/api_compatibility.md (hunk 2 above)
Comment (hawkinsp) [reply docs/api_compatibility.md]: Both done.

File: docs/api_compatibility.md (hunk 3 above)
Comment (hawkinsp) [reply docs/api_compatibility.md:7]: I left the deprecation period at three months with a comment that it is a lower bound and in practice deprecations may take longer. Also, nothing stops us from lengthening the period later. WDYT?

File: docs/api_compatibility.md (hunk 4 above)
Comment (hawkinsp) [reply docs/api_compatibility.md]: PEP565 indicates that one should issue `DeprecationWarning` if the target of the warning is other python developers. Am I misunderstanding it or is it incorrect in practice?

**Comment by hawkinsp** [comment]
PTAL.

File: docs/api_compatibility.md (hunk 3 above)
Comment (gnecula) [reply docs/api_compatibility.md:7]: LGTM

**Review by gnecula — APPROVED** [review]

**Review by jakevdp — APPROVED** [review]

**Pull request merged**